### PR TITLE
feat: monkey-patch MessagePort onproperties

### LINKED ISSUE #34581: Zone.js doesn't monkey patch MessagePort#onmessage
# 🐞 bug report

### Affected Package
 The issue is caused by package @angular/zone.js

### Is this a regression?

 No, I don't think so.

### Description

[MessagePort](https://developer.mozilla.org/en-US/docs/Web/API/MessagePort) is a native browser API used to asynchronously pass messages between different JS execution contexts, which my team uses for communicating across frames.

We discovered today that MessagePort is only partially monkey patched.  Because MessagePort extends from EventTarget, [zone.js](https://github.com/angular/angular/blob/master/packages/zone.js/STANDARD-APIS.md#browser) does correctly monkeypatch `MessagePort.prototype.addEventListener`, however it doesn't monkeypatch the `onmessage` property.

My team is using an internal package within google3 for iframe communication which uses `onmessage` instead of `addEventListener`. We could use `addEventListener` instead of `onmessage` however it's slightly cumbersome because then we need to call [MessagePort#start](https://developer.mozilla.org/en-US/docs/Web/API/MessagePort/start). `onmessage` implicitly calls MessagePort#start and I would guess most people using MessagePort would prefer to use `onmessage` because of this reason.

Given that `onmessage` can be monkey-patchable, it would be nice to fill in this hole in zone.js so other folks won't get confused. It took a while to debug this issue and at the very least, it would be good to [document](https://github.com/angular/angular/blob/master/packages/zone.js/STANDARD-APIS.md#browser) this limitation in zone.js (e.g. it only monkey-patches certain asynchronous APIs in MessagePort).

Thanks for your time.

## 🌍  Your Environment

**Angular Version:**
google3 cl/286903174

### PR DESCRIPTION
**Pull request by JiaLiPassion**
Close #34581

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #34581 

## What is the new behavior?
monkey-patch MessagePort.onmessage/onmessageerror properties. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/zone.js/test/browser/messageport.spec.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+fdescribe('MessagePort', () => {
```
Comment (alfaproject) [inline packages/zone.js/test/browser/messageport.spec.ts]: eff

Comment (JiaLiPassion) [reply packages/zone.js/test/browser/messageport.spec.ts]: thanks, forget to remove.

File: packages/zone.js/lib/browser/message-port.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+Zone.__load_patch('MessagePort', (global: any, Zone: ZoneType, api: _ZonePrivate) => {
```
Comment (mhevery) [inline packages/zone.js/lib/browser/message-port.ts]: please leave some space as it looks like the comment belongs to the function call.

File: packages/zone.js/test/browser/messageport.spec.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+describe('MessagePort', () => {
```
Comment (mhevery) [inline packages/zone.js/test/browser/messageport.spec.ts]: please leave some space as it looks like the comment belongs to the function call.

**Review by mhevery — APPROVED** [review]
Approved with some nits. 

Also the tests are failing, usually we try to get CI green before doing code review.

File: packages/zone.js/lib/browser/message-port.ts (hunk 2 above)
Comment (JiaLiPassion) [reply packages/zone.js/lib/browser/message-port.ts]: Got it, thanks, updated.

File: packages/zone.js/test/browser/messageport.spec.ts (hunk 3 above)
Comment (JiaLiPassion) [reply packages/zone.js/test/browser/messageport.spec.ts]: Got it, updated , thanks.

**Comment by JiaLiPassion** [comment]
@mhevery , thanks, the test failed because there is a flaky cases, I fixed it here, https://github.com/angular/angular/pull/35679

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:299418820:BASE:299418851:1583528752674:b6118229)

**Comment by AndrewKushnir** [comment]
FYI, the changes from this PR caused some failures in g3, so the changes were reverted from master. I'm re-opening this ticket.

cc @matsko

**Comment by JiaLiPassion** [comment]
@AndrewKushnir , could you tell me more information about the failure in g3, because this PR should have no impact for the current code, this PR just build a new `bundle` , and this bundle will not be loaded by default. Thanks!

**Comment by AndrewKushnir** [comment]
Hi @JiaLiPassion,

If I understand correctly, all files in the `browser` folder are included (except excluded ones), here is a snippet from the Angular BUILD file in g3:

```
ts_library(
    name = "zonejs_lib",
    srcs = glob(
        [
            "packages/zone.js/lib/browser/*.ts",
        ],
        exclude = [
            "packages/zone.js/lib/browser/rollup-*.ts",
            "packages/zone.js/lib/browser/shadydom.ts",
            "packages/zone.js/lib/browser/utils.ts",
            "packages/zone.js/lib/browser/webapis-*.ts",
        ],
    ) + [
        "rc/packages/zone.js/lib/zone-spec/long-stack-trace.ts",
        "rc/packages/zone.js/lib/zone-spec/wtf.ts",
    ],
    ...
)
```
So the newly added `packages/zone.js/lib/browser/message-port.ts` was also included and as a result affected some code.

Should it be in the `exclude` list?

Thank you.

**Comment by JiaLiPassion** [comment]
@AndrewKushnir , oh, yes, I see the code snippet also exclude `webapi-*` so this `messageport` should also be excluded. This is an optional packages, so I think `g3` doesn't need this file to be compiled. 
Thanks.

**Comment by AndrewKushnir** [comment]
Hi @JiaLiPassion, thanks for the update. Could you please rebase this PR, so that we re-run CI again and I'll try to run presubmit (excluding the file) in g3 again. Thank you.

**Comment by JiaLiPassion** [comment]
@AndrewKushnir , sure, I have rebased this PR, please review, thanks!

**Comment by AndrewKushnir** [comment]
@JiaLiPassion thanks for the update.

FYI, in order to run tests for ZoneJS changes that are in the merge queue, I've combined 4 ZoneJS PRs into one: https://github.com/angular/angular/pull/36043. Note: the https://github.com/angular/angular/pull/36043 PR is for testing purposes only and the changes will land in the corresponding PRs once g3 testing is completed.

Thank you.

**Comment by JiaLiPassion** [comment]
@AndrewKushnir, got it, thank you!

**Comment by AndrewKushnir** [comment]
FYI, changes in this PR were tested against g3 as a part of https://github.com/angular/angular/pull/36043, using Global TAP run, [which was successful](https://test.corp.google.com/ui#id=OCL:300603599:BASE:301077666:1584325797358:d8800cfc&include_skipped=true). Thank you.

**Pull request closed without merging**